ansible_runner/streaming.py

```diff
@@ -0,0 +1,214 @@
+                                     'job_events',
+                                     '{}-{}.json'.format(event_data['counter'],
+                                                         event_data['uuid']))
+
+        if self.event_handler is not None:
+            should_write = self.event_handler(event_data)
+        else:
+            should_write = True
+        for plugin in ansible_runner.plugins:
+            ansible_runner.plugins[plugin].event_handler(self.config, event_data)
+        if should_write:
+            with codecs.open(full_filename, 'w', encoding='utf-8') as write_file:
+                os.chmod(full_filename, stat.S_IRUSR | stat.S_IWUSR)
+                json.dump(event_data, write_file)
+
+    def artifacts_callback(self, artifacts_data):
+        buf = io.BytesIO(base64.b64decode(artifacts_data['payload']))
+        with zipfile.ZipFile(buf, 'r') as archive:
+            archive.extractall(path=self.config.artifact_dir)
+
+        if self.artifacts_handler is not None:
+            self.artifacts_handler(self.config.artifact_dir)
+
+
+class StreamWorker(object):
+    def __init__(self, worker_in, worker_out, **kwargs):
+        self.worker_in = worker_in
+        self.worker_out = worker_out
+
+        self.kwargs = kwargs
+
+        self.private_data_dir = tempfile.TemporaryDirectory().name
+
+    def run(self):
+        for line in self.worker_in:
+            data = json.loads(line)
+            if data.get('private_data_dir'):
+                buf = io.BytesIO(base64.b64decode(data['payload']))
+                with zipfile.ZipFile(buf, 'r') as archive:
+                    archive.extractall(path=self.private_data_dir)
+
+        kwargs_path = os.path.join(self.private_data_dir, 'kwargs')
+        if os.path.exists(kwargs_path):
+            with open(kwargs_path, "r") as kwf:
+                kwargs = json.load(kwf)
+            if not isinstance(kwargs, dict):
+                raise ValueError("Invalid kwargs data")
+        else:
+            kwargs = {}
+
+        self.kwargs.update(kwargs)
+
+        self.kwargs['quiet'] = True
+        self.kwargs['private_data_dir'] = self.private_data_dir
+        self.kwargs['status_handler'] = self.status_handler
+        self.kwargs['event_handler'] = self.event_handler
+        self.kwargs['artifacts_handler'] = self.artifacts_handler
+        self.kwargs['finished_callback'] = self.finished_callback
+
+        ansible_runner.interface.run(**self.kwargs)
+
+        # FIXME: do cleanup on the tempdir
+
+    def status_handler(self, status, runner_config):
+        self.worker_out.write(json.dumps(status).encode('utf-8'))
+        self.worker_out.write(b'\n')
+        self.worker_out.flush()
+
+    def event_handler(self, event_data):
+        self.worker_out.write(json.dumps(event_data).encode('utf-8'))
+        self.worker_out.write(b'\n')
+        self.worker_out.flush()
+
+    def artifacts_handler(self, artifact_dir):
```

Comment:
If we were to store a list of artifacts during the run, we could then bundle all of them as a single zip file after the end of the JSON events, and send a single "all artifacts" zip at the end of the process - again saving bandwidth if there are any artifacts including large binary data.

Reply:
In practice, I am sending all of the artifact files together as a single zip.  I used this callback/handler paradigm, though, since it felt like it fit well with the other callbacks.

Reply:
Oh, I get it, you're sending a JSON stub with `artifacts: True` and then you know to read a zip file from the input.  So if there _were_ intermingled job events and artifacts, it would still work.  Nicely done!

Reply:
Probably not any more, though, since now I'm unconditionally read()-ing after getting the `artifacts: True` blob and treating it as a single zip file.

Reply:
Now I'm doing `artifacts: <len>`, so the potential for intermingling is back on.